hb.elpi

```diff
@@ -1179,32 +1198,26 @@ declare-builder SrcFactory SrcFactoryRequires F FTy MiddleFactory TgtMixin FromI
   GoF = {mk-mixin-fun.then SrcFactoryRequires
           (factory-comp ML SrcFactory F MiddleFactory TgtMixin)},
 
-  std.assert-ok! (coq.typecheck GoF _GoFTy) "declare-builder: GoF illtyped",
+  std.spy(std.assert-ok! (coq.typecheck GoF _GoFTy) "declare-builder: GoF illtyped"),
   Name is {gref->modname SrcFactory} ^ "_to_" ^ {gref->modname TgtMixin} ^ {term_to_string {new_int}},
   hb-add-const Name GoF _GoFTy @transparent! GoFC,
   NewFrom = from SrcFactory TgtMixin (global (const GoFC)),
 ].
 
-% [main-declare-builder (builder _ F B) From MoreFrom] Given B of type FB, it
-% generates all the builders for mixins in FB from F via components of B.
-% From holds the (from F Mi Bi) predicates during folding.
+% [main-declare-builder (builder _ F M B) From MoreFrom] Given B of type FB, it
+% declares all the new builders F to M via B.
+% From holds the (from F Mi Bi) new clauses during folding.
 pred main-declare-builder i:builder, i:list prop, o:list prop.
-main-declare-builder (builder _ SrcFactory B) FromClauses MoreFromClauses :- !, std.do! [
+main-declare-builder (builder _ SrcFactory TgtMixin B) FromClauses MoreFromClauses :-
   FromClauses => std.do! [
-    std.assert-ok! (coq.typecheck B BTy) "main-declare-builder: B illtyped",
-    factory-alias->gref {target-gref BTy} MiddleFactory,
-    factory-requires SrcFactory SrcFactoryRequireswP,
-    list-w-params_list SrcFactoryRequireswP SrcFactoryRequires,
-
-    factory-provides MiddleFactory AllTgtMixinswP,
-    list-w-params_list AllTgtMixinswP AllTgtMixins,
-
-    std.filter AllTgtMixins (t\ not (std.mem! SrcFactoryRequires t)) TgtMixins,
-  ],
-  if-verbose (coq.say "HB: declare builders from" {nice-gref->string SrcFactory}
-                      "to" {std.map TgtMixins nice-gref->string}),
-  std.fold TgtMixins FromClauses (declare-builder SrcFactory SrcFactoryRequireswP B BTy MiddleFactory) MoreFromClauses,
-].
+    if (from SrcFactory TgtMixin _)
```

Comment:
Does this make sense? Or should we check something?

Reply:
We change the approach: hb.instance names all the mixin inhabitants and hb.end declares a `from` for them after section discharge.